blocks/ai-text/src/app/generate-text.tsx

```diff
@@ -0,0 +1,178 @@
+import { useGraphBlockModule } from "@blockprotocol/graph/react";
+import { useServiceBlockModule } from "@blockprotocol/service/react";
+import { FormEvent, useCallback, useEffect, useRef, useState } from "react";
+
+import { contentKey } from "../app";
+import { RootEntity } from "../types";
+import { TextPreview } from "./generate-text/text-preview";
+
+export const promptKey: keyof RootEntity["properties"] =
+  "https://blockprotocol.org/@blockprotocol/types/property-type/openai-text-model-prompt/";
+
+export const modelKey: keyof RootEntity["properties"] =
+  "https://blockprotocol.org/@blockprotocol/types/property-type/openai-text-model-name/";
+
+export const GenerateText = ({ blockEntity }: { blockEntity: RootEntity }) => {
+  const blockRootRef = useRef<HTMLDivElement>(null);
+  const inputRef = useRef<HTMLInputElement>(null);
+
+  const { graphModule } = useGraphBlockModule(blockRootRef);
+  const { serviceModule } = useServiceBlockModule(blockRootRef);
+
+  const [loading, setLoading] = useState(false);
+  const [errorMessage, setErrorMessage] = useState("");
+
+  // @todo implement model selection UI
+  // @see https://app.asana.com/0/1203358502199087/1203701786066059/f
+  const [model, _setModel] = useState("text-davinci-003");
+  const [promptText, setPromptText] = useState("");
+  const [generatedText, setGeneratedText] = useState("");
+
+  const {
+    metadata: {
+      entityTypeId,
+      recordId: { entityId },
+    },
+  } = blockEntity;
+
+  const onSubmit = useCallback(
+    async (event: FormEvent) => {
+      event.preventDefault();
+
+      if (loading || !promptText.trim()) {
+        return;
+      }
+
+      setErrorMessage("");
+      setLoading(true);
+
+      const { data, errors } = await serviceModule.openaiCompleteText({
+        data: {
+          max_tokens: 1000,
+          model,
+          prompt: promptText,
+        },
+      });
+
+      const textResponse = data?.choices[0]?.text;
+
+      if (errors || !textResponse) {
+        setErrorMessage("An error occurred");
+        setLoading(false);
+        return;
+      }
+
+      setGeneratedText(textResponse.replace(/^\n\n/, ""));
+
+      setLoading(false);
+      inputRef.current?.blur();
+    },
+    [loading, model, promptText, serviceModule],
+  );
+
+  useEffect(() => {
+    inputRef.current?.focus();
+  }, []);
+
+  const confirm = () =>
+    graphModule.updateEntity({
+      data: {
+        entityId,
+        entityTypeId,
+        properties: {
+          [contentKey]: generatedText,
+          [modelKey]: model,
+          [promptKey]: promptText,
+        },
+      },
+    });
```

Comment:
Nit: you could use `useCallback` here as well
```suggestion
  const handleConfirm = useCallback(
    () =>
      graphModule.updateEntity({
        data: {
          entityId,
          entityTypeId,
          properties: {
            [contentKey]: generatedText,
            [modelKey]: model,
            [promptKey]: promptText,
          },
        },
      }),
    [graphModule, entityId, entityTypeId, generatedText, model, promptText],
  );
```

Reply:
At the moment I don't think it's worth doing this – the dependency array is almost everything that might cause a re-render in this component anyway, so I expect you'd be constructing the function on every render anyway, except with an extra step of comparing these items before you did so.